core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,417 @@
+use crate::trie::POISONED_LOCK_ERR;
+use crate::{DBCol, StorageError, Store, TrieCache, TrieCachingStorage, TrieStorage};
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardUId;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::TrieNodesCount;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tracing::error;
+
+use super::trie_storage::BoundedQueue;
+
+const MAX_QUEUED_WORK_ITEMS: usize = 16 * 1024;
+const MAX_PREFETCH_STAGING_MEMORY: usize = 200 * 1024 * 1024;
+/// How much memory capacity is reserved for each prefetch request.
+/// Set to 4MiB, the same as `max_length_storage_value`.
+const PREFETCH_RESERVED_BYTES_PER_SLOT: usize = 4 * 1024 * 1024;
+
+/// Storage used by I/O threads to prefetch data.
+///
+/// This implements `TrieStorage` and therefore can be used inside a `Trie`.
+/// Prefetching runs through the normal trie lookup code and only the backing
+/// trie storage behaves differently.
+///
+/// `TriePrefetchingStorage` instances are always linked to a parent
+/// `TrieCachingStorage`. They share a shard cache, to avoid reading anything
+/// from the DB that is already cached.
+/// They communicate through `PrefetchStagingArea` exclusively.
+///
+/// Each I/O threads will have its own copy of `TriePrefetchingStorage`, so
+/// this should remain a cheap object.
+#[derive(Clone)]
+struct TriePrefetchingStorage {
+    /// Store is shared with parent `TrieCachingStorage`.
+    store: Store,
+    shard_uid: ShardUId,
+    /// Shard cache is shared with parent `TrieCachingStorage`. But the
+    /// pre-fetcher uses this in read-only mode to avoid premature evictions.
+    shard_cache: TrieCache,
+    /// Shared with parent `TrieCachingStorage`.
+    prefetching: Arc<Mutex<PrefetchStagingArea>>,
+}
+
+/// This type is shared between runtime crate and store crate.
+///
+/// The former creates `PrefetchApi` and puts requests in, the latter serves requests.
+/// With this API, the store does not know about receipts etc, and the runtime
+/// does not know about the trie structure. The only thing they share is this object.
+#[derive(Clone)]
+pub struct PrefetchApi {
+    /// Bounded, shared queue for all IO threads to take work from.
+    ///
+    /// Work items are defined as `TrieKey` because currently the only
+    /// work is to prefetch a trie key. If other IO work is added, consider
+    /// changing the queue to an enum.
+    work_queue: Arc<Mutex<BoundedQueue<TrieKey>>>,
```

Comment:
If I got it right, `BoundedQueue` is not useful here - if it is full, it pops and returns element from tail, instead of returning new element back. We may need different queue implementation here, sorry if naming was misleading

Reply:
Oh, I see, my mistake. I switched to the bounded queue last minute, doing manual checks over a VecDequeue should be simple enough anyway, I don't think I need a new queue type.

Reply:
probably I should add tests for the memory bounds too, this one should be caught in a test IMO